src/platforms/gbm-kms/server/linux_dmabuf.cpp

```diff
@@ -0,0 +1,846 @@
+    glGenTextures(1, &tex);
+    return tex;
+}
+
+bool drm_format_has_alpha(uint32_t format)
+{
+    /* TODO: We should really have something like libweston/pixel-formats.h
+     * We've said this multiple times before, so 🤷
+     */
+
+    switch (format)
+    {
+        /* This is only an optimisation, so pick a bunch of formats and hope that
+         * covers it…
+         */
+        case DRM_FORMAT_XBGR2101010:
+        case DRM_FORMAT_BGRX1010102:
+        case DRM_FORMAT_XBGR8888:
+        case DRM_FORMAT_BGRX8888:
+        case DRM_FORMAT_XRGB2101010:
+        case DRM_FORMAT_RGBX1010102:
+        case DRM_FORMAT_RGBX8888:
+        case DRM_FORMAT_XRGB8888:
+            return false;
+        default:
+        // TODO: I *think* true is a safe default, as it'll just mean we're blending something without alpha?
+            return true;
+    }
+}
+
+class WaylandDmabufTexBuffer :
+    public mg::BufferBasic,
+    public mg::NativeBufferBase,
+    public mg::gl::Texture
+{
+public:
+    // Note: Must be called with a current EGL context
+    WaylandDmabufTexBuffer(
+        DmaBufBuffer& source,
+        mg::EGLExtensions const& extensions,
+        std::shared_ptr<mir::renderer::gl::Context> ctx,
+        std::function<void()>&& on_consumed,
+        std::function<void()>&& on_release,
+        std::shared_ptr<mir::Executor> wayland_executor)
+        : ctx{std::move(ctx)},
+          tex{get_tex_id()},
+          on_consumed{std::move(on_consumed)},
+          on_release{std::move(on_release)},
+          size_{source.size()},
+          layout_{source.layout()},
+          has_alpha{drm_format_has_alpha(source.format())},
+          wayland_executor{std::move(wayland_executor)}
+    {
+        eglBindAPI(EGL_OPENGL_ES_API);
+
+        glBindTexture(GL_TEXTURE_2D, tex);
+        extensions.glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, source.reimport_egl_image());
+        // tex is now an EGLImage sibling, so we can free the EGLImage without
+        // freeing the backing data.
+
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    }
+
+    ~WaylandDmabufTexBuffer() override
+    {
+        wayland_executor->spawn(
+            [context = ctx, tex = tex]()
+            {
+              context->make_current();
+
+              glDeleteTextures(1, &tex);
+
+              context->release_current();
+            });
+
+        on_release();
+    }
+
+    std::shared_ptr<mir::graphics::NativeBuffer> native_buffer_handle() const override
+    {
+        return {nullptr};
+    }
+
+    mir::geometry::Size size() const override
+    {
+        return size_;
+    }
+
+    MirPixelFormat pixel_format() const override
+    {
+        // There's no way to implement this corectly…
```

Comment:
Is there a discussion to be had?

Reply:
Yes. We should lift something like [weston's pixel format descriptors](https://gitlab.freedesktop.org/wayland/weston/-/blob/master/libweston/pixel-formats.h) and replace the MirPixelFormat enum.

That's a change touching lots of code, though.